src/auth/jwt.strategy.ts

```diff
@@ -0,0 +1,28 @@
+import { Injectable, UnauthorizedException } from '@nestjs/common';
+import { PassportStrategy } from '@nestjs/passport';
+import { ExtractJwt, Strategy } from 'passport-jwt';
+import { Env } from 'src/env/index';
+import { JwtPayload } from './jwt-interface';
+import { RedisService } from '@liaoliaots/nestjs-redis';
+
+@Injectable()
+export class JwtStrategy extends PassportStrategy(Strategy, 'jwt') {
+  constructor(
+    private readonly redisService: RedisService,
+  ) {
+    super({
+      secretOrKey: Env.JWT.AUTH_TOKEN,
+      jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
+    });
+  }
+
+  async validate(payload: JwtPayload) {
+    const { userId } = payload;
+    const client = this.redisService.getClient();
+    const jwtToken = await client.get(`jwt:${userId}`);
+    if (!jwtToken) {
+      throw new UnauthorizedException();
```

Comment:
PullProofAI

There is a potential performance issue in the `JwtStrategy` class. The current implementation retrieves the JWT token from Redis synchronously, which can lead to blocking the event loop and impacting system performance, especially under high load.

#### Suggested Improvement

To address this performance concern, consider refactoring the code to use asynchronous operations when interacting with Redis. By utilizing async/await or promises, you can make the Redis operations non-blocking and improve the overall responsiveness of the application.

#### Proposed Code Change

```typescript
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { PassportStrategy } from '@nestjs/passport';
import { ExtractJwt, Strategy } from 'passport-jwt';
import { Env } from 'src/env/index';
import { JwtPayload } from './jwt-interface';
import { RedisService } from '@liaoliaots/nestjs-redis';

@Injectable()
export class JwtStrategy extends PassportStrategy(Strategy, 'jwt') {
  constructor(
    private readonly redisService: RedisService,
  ) {
    super({
      secretOrKey: Env.JWT.AUTH_TOKEN,
      jwtFromRequest: ExtractJwt.fromAuthHeaderAsBearerToken(),
    });
  }

  async validate(payload: JwtPayload) {
    const { userId } = payload;
    
    try {
      const client = this.redisService.getClient();
      const jwtToken = await client.get(`jwt:${userId}`);
      
      if (!jwtToken) {
        throw new UnauthorizedException();
      }
      
      return payload;
    } catch (error) {
      // Handle any errors here
      throw new UnauthorizedException();
    }
  }
}
```

By making these changes, the code will be more performant and responsive, ensuring that the application can handle concurrent requests efficiently without blocking the event loop.


<!-- This is an auto-generated comment by PullProofAI -->